Add RaceConditionDetector

disconnect3d opened the pull request:
For a contract of:
```
pragma solidity ^0.4.17;

contract Caller {
	address public stored_address;

	function callstoredaddress() {
	    stored_address.call();
	}	
    
	function setstoredaddress(address addr) {
	    stored_address = addr;
	}
}
```

It returns:
```
✘ 1 ➜ manticore --detect-transaction-order ./trans_order_dependence.sol
2018-09-14 14:58:45,557: [29299] m.main:INFO: Beginning analysis
2018-09-14 14:58:45,557: [29299] m.ethereum:INFO: Starting symbolic create contract
2018-09-14 14:58:45,868: [29299] m.ethereum:INFO: Starting symbolic transaction: 0
2018-09-14 14:58:49,621: [29299] m.ethereum:WARNING: Potential transaction order dependency:
Value has been stored in storage slot/index 0 in transaction that called setstoredaddress(address) and is now used in transaction that calls setstoredaddress(address).
An attacker seeing a transaction to setstoredaddress(address) could create a transaction to setstoredaddress(address) with high gas and win a race.
2018-09-14 14:58:55,299: [29299] m.ethereum:INFO: 3 alive states, 5 terminated states
2018-09-14 14:58:55,726: [29299] m.ethereum:INFO: Starting symbolic transaction: 1
2018-09-14 14:59:09,550: [29299] m.ethereum:WARNING: Potential transaction order dependency:
Value has been stored in storage slot/index 0 in transaction that called setstoredaddress(address) and is now used in transaction that calls stored_address().
An attacker seeing a transaction to stored_address() could create a transaction to setstoredaddress(address) with high gas and win a race.
2018-09-14 14:59:21,876: [29299] m.ethereum:WARNING: Potential transaction order dependency:
Value has been stored in storage slot/index 0 in transaction that called setstoredaddress(address) and is now used in transaction that calls callstoredaddress().
An attacker seeing a transaction to callstoredaddress() could create a transaction to setstoredaddress(address) with high gas and win a race.
2018-09-14 14:59:34,479: [29299] m.ethereum:INFO: 12 alive states, 20 terminated states
2018-09-14 14:59:35,013: [1090] m.ethereum:INFO: Generated testcase No. 0 - RETURN
2018-09-14 14:59:36,879: [1090] m.ethereum:INFO: Generated testcase No. 1 - STOP
2018-09-14 14:59:39,026: [1090] m.ethereum:INFO: Generated testcase No. 2 - RETURN
2018-09-14 14:59:41,186: [1090] m.ethereum:INFO: Generated testcase No. 3 - STOP
2018-09-14 14:59:43,093: [1090] m.ethereum:INFO: Generated testcase No. 4 - STOP
2018-09-14 14:59:45,129: [1090] m.ethereum:INFO: Generated testcase No. 5 - STOP
2018-09-14 14:59:46,995: [1090] m.ethereum:INFO: Generated testcase No. 6 - STOP
2018-09-14 14:59:48,884: [1090] m.ethereum:INFO: Generated testcase No. 7 - STOP
2018-09-14 14:59:51,096: [1090] m.ethereum:INFO: Generated testcase No. 8 - STOP
2018-09-14 14:59:52,978: [1090] m.ethereum:INFO: Generated testcase No. 9 - STOP
2018-09-14 14:59:55,311: [1090] m.ethereum:INFO: Generated testcase No. 10 - STOP
2018-09-14 14:59:57,400: [1090] m.ethereum:INFO: Generated testcase No. 11 - RETURN
2018-09-14 14:59:59,283: [1090] m.ethereum:INFO: Generated testcase No. 12 - STOP
2018-09-14 15:00:01,179: [1090] m.ethereum:INFO: Generated testcase No. 13 - REVERT
2018-09-14 15:00:02,996: [1090] m.ethereum:INFO: Generated testcase No. 14 - REVERT
2018-09-14 15:00:05,016: [1090] m.ethereum:INFO: Generated testcase No. 15 - REVERT
2018-09-14 15:00:07,325: [1090] m.ethereum:INFO: Generated testcase No. 16 - REVERT
2018-09-14 15:00:09,270: [1090] m.ethereum:INFO: Generated testcase No. 17 - REVERT
2018-09-14 15:00:10,563: [1090] m.ethereum:INFO: Generated testcase No. 18 - RETURN
2018-09-14 15:00:11,564: [1090] m.ethereum:INFO: Generated testcase No. 19 - REVERT
2018-09-14 15:00:13,443: [1090] m.ethereum:INFO: Generated testcase No. 20 - REVERT
2018-09-14 15:00:14,866: [1090] m.ethereum:INFO: Generated testcase No. 21 - REVERT
2018-09-14 15:00:16,817: [1090] m.ethereum:INFO: Generated testcase No. 22 - REVERT
2018-09-14 15:00:19,383: [1090] m.ethereum:INFO: Generated testcase No. 23 - REVERT
2018-09-14 15:00:21,277: [1090] m.ethereum:INFO: Generated testcase No. 24 - REVERT
2018-09-14 15:00:23,932: [1090] m.ethereum:INFO: Generated testcase No. 25 - REVERT
2018-09-14 15:00:25,263: [1090] m.ethereum:INFO: Generated testcase No. 26 - REVERT
2018-09-14 15:00:27,177: [1090] m.ethereum:INFO: Generated testcase No. 27 - REVERT
2018-09-14 15:00:29,619: [1090] m.ethereum:INFO: Generated testcase No. 28 - REVERT
2018-09-14 15:00:31,659: [1090] m.ethereum:INFO: Generated testcase No. 29 - STOP
2018-09-14 15:00:33,163: [1090] m.ethereum:INFO: Generated testcase No. 30 - RETURN
2018-09-14 15:00:35,258: [1090] m.ethereum:INFO: Generated testcase No. 31 - REVERT
2018-09-14 15:00:37,561: [29299] m.ethereum:INFO: Results in <somepath>/mcore_yueaswqd
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/trailofbits/manticore/1138)
<!-- Reviewable:end -->

feliam commented on manticore/ethereum.py:
```diff
@@ -161,6 +163,19 @@ class Detector(Plugin):
+        # So: if we are in the dispatcher, one of the bytes will effectively have more than one solutions
+
+        # TODO / FIXME: Benchmark/check if doing it byte by byte and also returning the bytes is faster
+        return len(state.solve_n(state.platform.current_transaction.data[:4], 2)) > 1
```
heuristic could be improved..
- this works when the tx input starts having several possible solutions. 
- not differentiated from the fallback func. 
- the result may be cached globally. (Address, pc) -> dispatcher/funcname/fallback

Uh.. if there is SolidityMetadata associated you could use that too.

disconnect3d replied:
@feliam 
> this works when the tx input starts having several possible solutions.
Hmm... So probably the only edge case is situation where contract have no other functions than a fallback one which and with current approach we won't detect dispatcher properly, is that right?

Do you have any idea how to distinguish `fallback` vs `dispatch`?

> the result may be cached globally. (Address, pc) -> dispatcher/funcname/fallback

Do you think it is okay to map `(address,pc)` to `FuncInfo` such as:
```python
class FuncType:
    DISPATCHER = 0
    FALLBACK = 1
    USER_FUNC = 2
    # maybe also constructor? but its not really used in transactions, so maybe not?

FuncInfo = namedtuple('FuncInfo', 'type, signature')
```

feliam replied:
Idea to differentiate? Not much.
- If you have access to the metadata you can try to parse the output from an  metadata.get_source_for(evm_offset).
- Use the compiler in other way
- Number of instructions executed could help in the heuristics otherwise 

As for the mapping to functype look out for the runtime vs. init code. Maybe you need to add a flag there to decide constructor/runtime. Other than that whatever works. but do it on the global manticore.context dict.

disconnect3d commented:
I've rebased the branch, changed name to `RaceConditionDetector`, added a warning to it and better description.

Here is an example output for the example contract above:
```
(manticore) dc@ubuntu:~$ manticore --detect-race-condition ./vuln.sol
/home/dc/manticore_project/manticore/ethereum/detectors.py:599: UserWarning: The RaceCondition detector might not work properly for contracts that have only a fallback function. See the detector's implementation and it's `_is_in_dispatcher` method for more information.
  "The RaceCondition detector might not work properly for contracts that have only a fallback function. "
2018-10-24 05:12:56,057: [17216] m.main:INFO: Beginning analysis
2018-10-24 05:12:56,058: [17216] m.ethereum:INFO: Starting symbolic create contract
2018-10-24 05:12:56,421: [17216] m.ethereum:INFO: Starting symbolic transaction: 0
2018-10-24 05:13:05,476: [17216] m.ethereum:INFO: 3 alive states, 5 terminated states
2018-10-24 05:13:05,834: [17216] m.ethereum:INFO: Starting symbolic transaction: 1
2018-10-24 05:13:19,224: [17216] m.e.detectors:WARNING: Potential race condition (transaction order dependency):
Value has been stored in storage slot/index 0 in transaction that called setstoredaddress(address) and is now used in transaction that calls stored_address().
An attacker seeing a transaction to stored_address() could create a transaction to setstoredaddress(address) with high gas and win a race.
2018-10-24 05:13:25,311: [17216] m.e.detectors:WARNING: Potential race condition (transaction order dependency):
Value has been stored in storage slot/index 0 in transaction that called setstoredaddress(address) and is now used in transaction that calls setstoredaddress(address).
An attacker seeing a transaction to setstoredaddress(address) could create a transaction to setstoredaddress(address) with high gas and win a race.
2018-10-24 05:13:29,649: [17216] m.e.detectors:WARNING: Potential race condition (transaction order dependency):
Value has been stored in storage slot/index 0 in transaction that called setstoredaddress(address) and is now used in transaction that calls callstoredaddress().
An attacker seeing a transaction to callstoredaddress() could create a transaction to setstoredaddress(address) with high gas and win a race.
2018-10-24 05:13:40,538: [17216] m.ethereum:INFO: 12 alive states, 20 terminated states
2018-10-24 05:13:40,922: [21548] m.ethereum:INFO: Generated testcase No. 0 - RETURN
2018-10-24 05:13:42,508: [21548] m.ethereum:INFO: Generated testcase No. 1 - STOP
2018-10-24 05:13:43,730: [21548] m.ethereum:INFO: Generated testcase No. 2 - RETURN
2018-10-24 05:13:45,330: [21548] m.ethereum:INFO: Generated testcase No. 3 - STOP
2018-10-24 05:13:46,773: [21548] m.ethereum:INFO: Generated testcase No. 4 - STOP
2018-10-24 05:13:48,110: [21548] m.ethereum:INFO: Generated testcase No. 5 - STOP
2018-10-24 05:13:49,855: [21548] m.ethereum:INFO: Generated testcase No. 6 - STOP
2018-10-24 05:13:51,308: [21548] m.ethereum:INFO: Generated testcase No. 7 - STOP
2018-10-24 05:13:52,923: [21548] m.ethereum:INFO: Generated testcase No. 8 - STOP
2018-10-24 05:13:54,900: [21548] m.ethereum:INFO: Generated testcase No. 9 - STOP
2018-10-24 05:13:56,634: [21548] m.ethereum:INFO: Generated testcase No. 10 - STOP
2018-10-24 05:13:58,015: [21548] m.ethereum:INFO: Generated testcase No. 11 - RETURN
2018-10-24 05:13:59,252: [21548] m.ethereum:INFO: Generated testcase No. 12 - STOP
2018-10-24 05:14:00,710: [21548] m.ethereum:INFO: Generated testcase No. 13 - REVERT
2018-10-24 05:14:02,085: [21548] m.ethereum:INFO: Generated testcase No. 14 - REVERT
2018-10-24 05:14:03,345: [21548] m.ethereum:INFO: Generated testcase No. 15 - REVERT
2018-10-24 05:14:04,729: [21548] m.ethereum:INFO: Generated testcase No. 16 - REVERT
2018-10-24 05:14:06,499: [21548] m.ethereum:INFO: Generated testcase No. 17 - REVERT
2018-10-24 05:14:07,473: [21548] m.ethereum:INFO: Generated testcase No. 18 - RETURN
2018-10-24 05:14:08,208: [21548] m.ethereum:INFO: Generated testcase No. 19 - REVERT
2018-10-24 05:14:09,789: [21548] m.ethereum:INFO: Generated testcase No. 20 - REVERT
2018-10-24 05:14:10,792: [21548] m.ethereum:INFO: Generated testcase No. 21 - REVERT
2018-10-24 05:14:12,177: [21548] m.ethereum:INFO: Generated testcase No. 22 - REVERT
2018-10-24 05:14:13,613: [21548] m.ethereum:INFO: Generated testcase No. 23 - REVERT
2018-10-24 05:14:15,403: [21548] m.ethereum:INFO: Generated testcase No. 24 - REVERT
2018-10-24 05:14:16,749: [21548] m.ethereum:INFO: Generated testcase No. 25 - REVERT
2018-10-24 05:14:17,834: [21548] m.ethereum:INFO: Generated testcase No. 26 - REVERT
2018-10-24 05:14:19,858: [21548] m.ethereum:INFO: Generated testcase No. 27 - REVERT
2018-10-24 05:14:21,225: [21548] m.ethereum:INFO: Generated testcase No. 28 - REVERT
2018-10-24 05:14:22,655: [21548] m.ethereum:INFO: Generated testcase No. 29 - STOP
2018-10-24 05:14:23,763: [21548] m.ethereum:INFO: Generated testcase No. 30 - RETURN
2018-10-24 05:14:24,839: [21548] m.ethereum:INFO: Generated testcase No. 31 - REVERT
2018-10-24 05:14:27,516: [17216] m.ethereum:INFO: Results in /home/dc/mcore_urujglxr
```

@mossberg @feliam pls review

feliam commented on manticore/ethereum/detectors.py:
```diff
@@ -577,3 +577,111 @@ def did_evm_read_storage_callback(self, state, address, offset, value):
+        super().__init__(*a, **kw)
+
+    @staticmethod
+    def _is_in_dispatcher(state):
```
You could cache this. Save this as a flag in state.context['in_dispatcher'] = {{None, True, False}}
So you do not need to re-solve once you already know.

Also the fallback and the dispatcher part are kind of indistinguishable this way. 
The fallback function is reached with the 4bytes beign diff from the known/defined functions. So at the fallback the 4bytes will have more than 2 solutions but never any of the unselected funcs.

disconnect3d replied:
> Also the fallback and the dispatcher part are kind of indistinguishable this way.

Hmm... right. I think I will rename it to `_in_user_func` and make it return `return len(state.solve_n(state.platform.current_transaction.data[:4], 2)) == 1` instead.

mossberg commented on manticore/core/plugin.py:
```diff
@@ -33,7 +34,8 @@ def context(self):
 
     def __init__(self):
-        self.manticore = None
+        from manticore import Manticore
```
this import is necessary for the type annotation?

disconnect3d replied:
Yes

mossberg replied:
ok. it strikes me a little odd to put an import here for the sole purpose being able to use the type annotation... ? also, i feel like the type of this attribute is pretty clear, given it's name. so maybe we don't need it?

disconnect3d replied:
It was just for IDE purposes. I am okay with removing it.

disconnect3d commented on manticore/ethereum/detectors.py:
```diff
@@ -577,3 +577,112 @@ def did_evm_read_storage_callback(self, state, address, offset, value):
+        """
+
+        # TODO / FIXME: Benchmark/check if doing it byte by byte and also returning the bytes is faster
+        return len(state.solve_n(state.platform.current_transaction.data[:4], 2)) == 1
```
@feliam we were speaking about caching this result; what would be the proper cache key? Current pc? Is that okay to take a hash out of it then as it may be e.g. a `BitVecAdd`?

disconnect3d commented on tests/eth_detectors.py:
```diff
@@ -57,10 +54,11 @@ def _test(self, name, should_find, use_ctor_sym_arg=False):
 
-        expected_findings = set(((c, d) for b, c, d in should_find))
-        actual_findings = set(((c, d) for a, b, c, d in mevm.global_findings))
+        expected_findings = set(((finding, at_init) for _pc, finding, at_init in should_find))
```
Shall we keep `pc` in `should_find`? It seems it is redundant..

feliam commented on manticore/__main__.py:
```diff
@@ -130,6 +130,10 @@ def positive(value):
     eth_flags.add_argument('--avoid-constant', action='store_true',
                            help='Avoid exploring constant functions for human transactions')
 
+    parser.add_argument('--detect-race-condition', action='store_true',
```
Did not this changed in favor of that eth_flags ?
